### ISSUE TITLE
A setter is not generated in class ScriptProcessorNode

### ISSUE THREAD

**Issue by Larpoux** [fzyzcjy/flutter_rust_bridge#2167]
### Describe the bug

In ```class PannerNode``` there is a simple getter not generated: 
```rust
#[derive(Debug)]
pub struct StereoPannerNode {
    /// Represents the node instance and its associated audio context
    registration: AudioContextRegistration,
    /// Infos about audio node channel configuration
    channel_config: ChannelConfig,
    /// The position of the input in the output’s stereo image. -1 represents
    /// full left, +1 represents full right.
    pan: AudioParam,
}

impl AudioNode for StereoPannerNode {
    ...
    /// Returns the pan audio parameter
    #[must_use]
    pub fn pan(&self) -> &AudioParam {
        &self.pan
    }
}
```

### Steps to reproduce

-

### Logs

```shell
-
```

**Comment by fzyzcjy** [comment]
Hmm, the frb_example one does generate this, maybe try to fully copy code there.

Code directly related to this: The macro call below.

(p.s. Macros in rust is similar to C++, but less error-prone)

![image](https://github.com/fzyzcjy/flutter_rust_bridge/assets/5236035/22639dbf-3e27-4e9a-a0f0-d7689467cc51)

**Comment by Larpoux** [comment]
OK, this is perfect. It was my fault : the W3C recommandation defined two class : 
- PannerNode
- StereoPannerNode

I used `PannerNode` instead for `stereoPannerNode`.
Everything OK

**Comment by Larpoux** [comment]
BTW : we will discuss one day to know if `override_web_auiod_api.rs` and `third_party/web_audio_api/*.rs` 
are part of the frb project or part of the τ project.
My actual feeling is that those files belong to τ and not frb.
But actually you are the only people on earth able to maintain them. [This is something which frighten me : if one day you stop to develop the τ surpport I will be an orphan].

**Comment by Larpoux** [comment]
It seems (but I can be wrong) that the following setter is not generated in class `ScriptProcessorNode` : 
```rust
pub fn [set_onaudioprocess](https://docs.rs/web-audio-api/latest/web_audio_api/node/struct.ScriptProcessorNode.html#method.set_onaudioprocess)<F: [FnMut](https://doc.rust-lang.org/nightly/core/ops/function/trait.FnMut.html)([AudioProcessingEvent](https://docs.rs/web-audio-api/latest/web_audio_api/struct.AudioProcessingEvent.html)) + [Send](https://doc.rust-lang.org/nightly/core/marker/trait.Send.html) + 'static>(
    &self,
    callback: F,
)
```

**Comment by Larpoux** [comment]
Another function missing in `AnalyserNode` : 
```rust
pub fn get_byte_time_domain_data(&mut self, buffer: &mut [u8])
```
```rust
    #[frb(ignore)]
    pub fn get_byte_time_domain_data() {}
```

**Comment by fzyzcjy** [comment]
> Everything OK

Looks great!

> BTW : we will discuss one day to know if override_web_auiod_api.rs and third_party/web_audio_api/*.rs
are part of the frb project or part of the τ project.
My actual feeling is that those files belong to τ and not frb.

Surely it is part of the audio thing instead of frb thing. It is inside frb simply because I need to have as many as possible tests which auto run on each commit via CI, in order to ensure things are working well. Thus, feel free to PR your changes back to frb example to have more tests!

> But actually you are the only people on earth able to maintain them. [This is something which frighten me : if one day you stop to develop the τ surpport I will be an orphan].

The code there simply follows https://cjycode.com/flutter_rust_bridge/guides/third-party/automatic + some Rust familiarity. So no worries - it is definitely impossible I am the *only* one knowing it! And again no worries, I will keep it maintained.

**Comment by fzyzcjy** [comment]
> It seems (but I can be wrong) that the following setter is not generated in class ScriptProcessorNode :

I guess it is because the complex generics confuses frb. One way is to mimic how other `set_something` is handled. For example, set_on_ended, set_on_processor_error, set_on_complete, ...

**Comment by fzyzcjy** [comment]
> Another function missing in AnalyserNode :

It's because `&mut [u8]` is not implemented yet. A quick workaround (performance is suboptimal) is:

```rust
#[ext]
pub impl AnalyserNode {
  pub fn frb_override_get_byte_time_domain_data(&mut self) -> Vec<u8> {
    let mut data = vec![0; arr_len_here];
    self.get_byte_time_domain_data(&mut data);
    data
  }
}
```

**Comment by Larpoux** [comment]
> It's because `&mut [u8]` is not implemented yet. A quick workaround (performance is suboptimal since it makes 1 allocation) is:

This is no problem for me : we will have one allocation more in rust, but we will have one allocation less in dart.

> One way is to mimic how other `set_something` is handled. For example, set_on_ended, set_on_processor_error, set_on_complete, ...

This would be OK, I guess : I don't understand why this function has a so weird signature : actually I just want to register a simple callback as I did with `set_on_complete`

**Comment by Larpoux** [comment]
In class `AudioBuffer` : 
```rust
    pub fn get_channel_data(&self, channel_number: usize) -> &[f32]
```
and
```rust
   pub fn get_channel_data_mut(&mut self, channel_number: usize) -> &mut [f32]
```
are generated as functions without results.
```dart

  /// Return a read-only copy of the underlying data of the channel
  ///
  /// # Panics
  ///
  /// This function will panic if:
  /// - the given channel number is greater than or equal to the given number of channels.
  Future<void> getChannelData({required int channelNumber});

  /// Return a mutable slice of the underlying data of the channel
  ///
  /// # Panics
  ///
  /// This function will panic if:
  /// - the given channel number is greater than or equal to the given number of channels.
  Future<void> getChannelDataMut({required int channelNumber});
```

**Comment by fzyzcjy** [comment]
As is mentioned in doc, frb does not support `&[sth]`. Quick workaround (again has more alloc):

```rust
  pub fn frb_override_get_channel_data(&self, channel_number: usize) -> Vec<f32> {
	self.get_channel_data().to_owned()
  }

  pub fn set_channel_data(&self, channel_number: usize, data: Vec<f32>) {
	self.get_channel_data_mut().copy_from_slice(&data)
  }
```

And `#[frb(ignore)]` the get_channel_data_mut since `&mut [sth]` is also not supported yet.

**Comment by Larpoux** [comment]
The problem is that `getChannelDataMut` returns a pointer to the buffer inside AudioBuffer so that the caller can modify the Audio Buffer data.
But I am not blocked on this problem. I can port the Mozilla examples under flutter without calling those fucking functions.
My real problem is actually : 
```rust
pub fn pub fn set_onaudioprocess<F: FnMut(AudioProcessingEvent) + Send + 'static>(
        &self,
        mut callback: F,
    )
    callback: F,
)
```
I am going to try to do same as you did for set_onend<...> because what you did was fine.

**Comment by Larpoux** [comment]
I successfully port a second Mozilla Javascript example (out of a total of 24 ex.) to τ. (It was a very simple one).

![MozillaEx_AudioBufferSource_graph](https://github.com/fzyzcjy/flutter_rust_bridge/assets/45900255/00647033-fe2a-442c-81dc-58a19c719b0b)

**Comment by fzyzcjy** [comment]
> The problem is that getChannelDataMut returns a pointer to the buffer inside AudioBuffer so that the caller can modify the Audio Buffer data.

Yes, that's why I workaround by providing a setter function.

> I am going to try to do same as you did for set_onend<...> because what you did was fine.

Ping me if there are any questions!

> I successfully port a second Mozilla Javascript example (out of a total of 24 ex.) to τ. (It was a very simple one).

Looks good!

That looks like a big workload to port 24 examples...

**Comment by Larpoux** [comment]
Please help 👎 
I need help with Rust : I spent several hours trying to find myself the solution, but Rust is too difficult for me : 
```rust
#[ext]
pub impl ScriptProcessorNode
{
    // NOTE: The original name was `set_onaudioprocess` and here the new name has `_`
    fn set_on_audio_process
    (
        &self,
        callback: impl Fn(AudioProcessingEvent) -> DartFnFuture<()> + Send + 'static,
    )
    {
        self.set_onaudioprocess(Box::new(|event|
        {
            FLUTTER_RUST_BRIDGE_HANDLER.async_runtime().spawn(async move  { callback(event).await });
        }))
    }
}
```
I get this error : 
```

error[E0525]: expected a closure that implements the `FnMut` trait, but this closure only implements `FnOnce`
   --> src/api/override_web_audio_api.rs:151:42
    |
151 |           self.set_onaudioprocess(Box::new(|event|
    |                ------------------ -        ^^^^^^^ this closure implements `FnOnce`, not `FnMut`
    |  ______________|__________________|
    | |              |
    | |              required by a bound introduced by this call
152 | |         {
153 | |             //FLUTTER_RUST_BRIDGE_HANDLER.async_runtime().spawn(callback(event));
154 | |             FLUTTER_RUST_BRIDGE_HANDLER.async_runtime().spawn(async move  { callback(event).await });
    | |                                                                             -------- closure is `FnOnce` because it moves the variable `callback` out of its environment
155 | |             //async move { callback(event).await }
156 | |         }))
    | |__________- the requirement to implement `FnMut` derives from here
    |
    = note: required for `Box<{closure@src/api/override_web_audio_api.rs:151:42: 151:49}>` to implement `FnMut<(web_audio_api::AudioProcessingEvent,)>`
note: required by a bound in `web_audio_api::node::ScriptProcessorNode::set_onaudioprocess`
   --> /Volumes/mac-J/larpoux/.cargo/registry/src/index.crates.io-6f17d22bba15001f/web-audio-api-1.0.0-rc.6/src/node/script_processor.rs:145:34
    |
145 |     pub fn set_onaudioprocess<F: FnMut(AudioProcessingEvent) + Send + 'static>(
    |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `ScriptProcessorNode::set_onaudioprocess`

For more information about this error, try `rustc --explain E0525`.
warning: `tau` (lib) generated 6 warnings
error: could not compile `tau` (lib) due to 1 previous error; 6 warnings emitted
```

**Comment by fzyzcjy** [comment]
Try to mimic set_on_state_change, which uses Arc to avoid the error that objects are moved

**Comment by Larpoux** [comment]
Thank you @fzyzcjy : it works fine 😄 

![MozillaEx_AudioAnalyser_preview](https://github.com/fzyzcjy/flutter_rust_bridge/assets/45900255/a0ea001d-87d7-40e7-ba0d-471206f18a1f)
![MozillaEx_AudioAnalyser_graph](https://github.com/fzyzcjy/flutter_rust_bridge/assets/45900255/7d7a56c6-1657-443f-9ab8-0659e10d2d4c)

**Comment by fzyzcjy** [comment]
🎉

**Comment by Larpoux** [comment]
Fixed by PR [#2187]. I hope that it will not be rejected.
